drivers/gpu/drm/meson/meson_cache_control.c

```diff
@@ -69,53 +70,79 @@ static void meson_drm_ump_osk_msync(struct drm_gem_cma_object *cma_obj, void *vi
 		}
 	}
 
-	if (!cma_obj)
+	if (!gem_obj)
 		return;
-	gem_obj = &cma_obj->base;
 
 	DBG_MSG(3, ("meson_drm_ump_osk_msync(): Flushing CPU L2 Cache\n"));
 
-	/* Flush L2 using physical addresses
-	 * Our allocations are always contiguous (GEM CMA), so we have only one mem block */
-
-	if (offset >= gem_obj->size) {
-		offset -= gem_obj->size;
-		return;
-	}
-
-	if (offset) {
-		start_p = (u32)cma_obj->paddr + offset;
-		/* We'll zero the offset later, after using it to calculate end_p. */
+	if (gem_obj->is_scattered) {
+		struct meson_drm_gem_scattered_object *meson_gem = to_meson_drm_gem_scattered_obj(gem_obj);
+		struct scatterlist *sgl;
+		int i;
+
+		for_each_sg(meson_gem->sgt->sgl, sgl, meson_gem->sgt->nents, i) {
+			// start_p = (u32)__pa(page_address(sg_page(sgl)));
+			start_p = sg_phys(sgl);
+			end_p = start_p + sgl->length;
+
+			switch (op) {
+			case DRM_MESON_MSYNC_CLEAN:
+				outer_clean_range(start_p, end_p);
+				break;
+			case DRM_MESON_MSYNC_CLEAN_AND_INVALIDATE:
+				outer_flush_range(start_p, end_p);
+				break;
+			case DRM_MESON_MSYNC_INVALIDATE:
+				outer_inv_range(start_p, end_p);
+				break;
+			default:
+				break;
+			}
+		}
 	} else {
-		start_p = (u32)cma_obj->paddr;
-	}
+		struct drm_gem_cma_object *cma_obj = to_drm_gem_cma_obj(gem_obj);
+		/* Flush L2 using physical addresses
+		 * Our allocations are always contiguous (GEM CMA), so we have only one mem block */
+
+		if (offset >= gem_obj->size) {
+			offset -= gem_obj->size;
+			return;
+		}
 
-	if (size < gem_obj->size - offset) {
-		end_p = start_p + size;
-		size = 0;
-	} else {
 		if (offset) {
-			end_p = start_p + (gem_obj->size - offset);
-			size -= gem_obj->size - offset;
-			offset = 0;
+			start_p = (u32)cma_obj->paddr + offset;
+			/* We'll zero the offset later, after using it to calculate end_p. */
 		} else {
-			end_p = start_p + gem_obj->size;
-			size -= gem_obj->size;
+			start_p = (u32)cma_obj->paddr;
+		}
+
+		if (size < gem_obj->size - offset) {
+			end_p = start_p + size;
+			size = 0;
+		} else {
+			if (offset) {
+				end_p = start_p + (gem_obj->size - offset);
+				size -= gem_obj->size - offset;
+				offset = 0;
+			} else {
+				end_p = start_p + gem_obj->size;
+				size -= gem_obj->size;
+			}
 		}
-	}
 
-	switch (op) {
-	case DRM_MESON_MSYNC_CLEAN:
-		outer_clean_range(start_p, end_p);
-		break;
-	case DRM_MESON_MSYNC_CLEAN_AND_INVALIDATE:
-		outer_flush_range(start_p, end_p);
-		break;
-	case DRM_MESON_MSYNC_INVALIDATE:
-		outer_inv_range(start_p, end_p);
-		break;
-	default:
-		break;
+		switch (op) {
+		case DRM_MESON_MSYNC_CLEAN:
```

Comment:
Can  you put this shared bit in a function or similar? `cache_control_op(op, start_p, end_p);`